java/sage/VideoFrame.java

```diff
@@ -5650,6 +5651,40 @@ private int getMatchingLangIndex(String[] langOptions, String targetLang)
 
   private boolean selectDefaultSubpicLanguage()
   {
+	String defaultAudioLang = uiMgr.get("default_audio_language", "English"); 
+    MediaLangInfo mli = (MediaLangInfo) mediaLangMap.get(defaultAudioLang);
+    SubpictureFormat [] subs = this.currFile.getFileFormat().getSubpictureFormats();
+
+	/*
+	 * Loop thru all subtitle tracks of the mediafile to see if there is a forced subtitle 
+	 * track that matches the configured default audio language.  If it does it will 
+	 * select the track for playback.
+    */	 
+    for(int i = 0; i < subs.length; i++)
+    {
+      if(subs[i].getForced())      
+      {
+        //Check for a match on the two letter lang code
+        if(subs[i].getLanguage().equalsIgnoreCase(mli.twoChar))
```

Comment:
You need to handle the case where mli is null.

Reply:
Actually, why aren't you calling performLanguageMatch with a modified subpicture array? (i.e. same length so the indices match, but then put empty strings for the non-forced subs)

Reply:
`
public String[] getDVDAvailableSubpictures()
  {
    sage.media.sub.SubtitleHandler subby = subHandler;
    if (subby != null && !subby.mpControlled())
    {
      return subby.getLanguages();
    }
    MediaPlayer mp = player;
    if (mp instanceof DVDMediaPlayer)
      return ((DVDMediaPlayer) mp).getDVDAvailableSubpictures();
    else
      return Pooler.EMPTY_STRING_ARRAY;
  }
`
Should I be concerned with what this code is doing?  I was having some difficulty in trying to figure out what subtitle tracks you are pulling back.  I assume from this code that depending on what player is active you may ask the player vs using the info from the format detection.  Am I correct in that assumption?  Should I only do my logic in certain cases?

When is the SubtitleHandler used vs the DVDMediaPlayer?  Should I only do my logic if using Subtitle Handler?

Reply:
It's not really about what player is active, it's about whether the subtitle tracks are stored internally with the file itself or are external files in the same directory. However, looking at the code...I can't figure out why it was done this way because they end up calling the same thing on the ContainerFormat in the end (but I'm pretty sure there's another subtle difference I'm forgetting).  That's why I'd recommend making this as aligned with how the current code operates rather than trying to do your own logic for figuring out what the track options are.

Reply:
@Narflex 

I agree with your comments, I just want to make sure I take the right approach on getting the subtitle tracks.

Should I continue to just use this.currFile.getFileFormat().getSubpictureFormats(); to get the subtitle tracks, or do I need to reverse engineer, and possibly make modifications to SubtitleHandler and all MediaPlayers that implement DVDMediaPlayer to make new methods that return forcedSubtitle tracks?

I took the easy approach because I was having difficulty tracing all of the places that Sage is getting subtitle tracks.  And this change only really impacts Subtitles inside the media file.

Josh

Reply:
There's no reason you need to change SubtitleHandler or any of the MediaPlayer code, it's outside the scope of what you're doing here.  The approach I would take is to modify ContainerFormat.getSubpictureStreamSelectionDescriptors so that it also appends onto the end " [Forced]" for forced sub tracks.  Then in VideoFrame.selectDefaultSubpicLanguage I would make a copy of the array of subpic tracks from getDVDAvailableSubpictures(), go through that array and set to the empty string all elements that don't have the "[Forced]" string in it.  Then call performLanguageMatch(defaultAudioLang, modifiedSubpicArray, DVD_CONTROL_SUBTITLE_CHANGE, getDVDSubpicture()) and if that doesn't return true, then proceed with normal default subpic selection....you could also skip that step if you don't find any Forced subs in that list.